nvaccess/nvda
# Simplify appveyor.yml

### Link to issue number:

None

### Summary of the issue:

If a developer wishes to use multiple appveyor.yml files, there would be a lot of code duplication. In order to make changes to the AppVeyor build process(es), inline powershell scripts must be edited. Inline powershell scripts are harder to lint and parse than individual files within IDEs. 

### Description of how this pull request fixes the issue:

Simplifies the `appveyor.yml` file by putting the powershell scripts into separate files. 

Note to reviewers: I would recommend reading through these changes, and then compare them to the individual commits and README. The git bash script is a tool you can use to review the final changes or specific commits. I've noted some proposed changes of behaviour, but otherwise original behaviour is intended.

Changes: 
- Moved comments to `appveyor\README.md` and created an outline of AppVeyor proccess in the `appveyor\README.md`
- Move inline ps into script files
  - adds `$ErrorActionPreference = "Stop";` to the start of each except our test scripts which are allowed to fail
  - confirm with the following (git) bash script, run from the project root. This generates diff files optimised for review between masters appveyor.yml and our powershell scripts on `HEAD`. To see the diff at a specific commit, check it out, as the files specified by `ls` must be checked out locally.
  ```sh
  # the git diff argument -W ensures whole functions are preserved in the diff, to aid reviewing and give context
  # the git diff arguments -wb ignore all trailing and leading whitespace changes. this is ignored due to
  # indentation changes.
  #
  # perl (as it supports advanced regex - PCRE) replaces our large blocks of removed code with a snippet
  # the perl preserves the first, second, second last, and last lines of blocks of removed code longer than
  # 4 lines. This makes it easier to diff between the relevant functions and preserves context (such as the
  # appveyor.yml build phase we are in).
  export subLine='\n-[^\n]*';  # subtracted line regex
  export perlSed="s/($subLine)($subLine)(?:$subLine)+($subLine)($subLine)/\1\2\n-...\3\4/g";
  for file in $(ls appveyor/scripts/*.ps1 appveyor/scripts/tests/*.ps1)
  do
  git diff -Wwb master:appveyor.yml HEAD:$file | perl -0777 -pe $perlSed > $file-appveyor.yml-hide-removed.diff
  done
  ```
- as the repo isn't cloned until after the `init` phase, `setBuildVersionVars.ps1` was moved until after that, into the `install` build phase
  - this means we don't update the appveyor build number until after a successful clone
- moved the translation comments check to the `test_script` build phase
- moved some inline cmd to ps and cleaned up appveyor.yml
- removed unnecessary exit commands
  - as any exit code above 0 (eg terminating error) will cause AppVeyor to exit and the build to fail - [Proof: see this failed build](https://github.com/nvaccess/nvda/commit/72cdb74adf3324ed7bf14ddc5cd4f3191753300f)
- removed markdown from the build link message as AppVeyor and other applications don't support it - only GitHub

### Testing strategy:

- Ensure each part and type of build occurs successfully. 
- See review tools to ensure no unintended changes of behaviour: especially `decryptFilesForSigning.ps1` as this acts upon secure files when on non PR builds. 
- Confirm intended changes of behaviour
- Check build artifacts are there and the same as other PR builds (and eventual master builds). 

### Known issues with pull request:

None needed

### Change log entries:

None needed

### Code Review Checklist:

  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
